src/Model/ModelData.php

```diff
@@ -487,6 +489,31 @@ public function hasValue(string $field, array $arguments = [], bool $cache = tru
 
     // UTILITY METHODS -------------------------------------------------------------------------------------------------
 
+    /**
+     * Flags provides the user with additional data about the current page status.
+     *
+     * Mostly this is used for versioning, but can be used for other purposes (e.g. localisation).
+     * Each page can have more than one status flag.
+     *
+     * Returns an associative array of a unique key to a (localized) title for the flag.
+     * The unique key can be reused as a CSS class.
+     *
+     * Example (simple):
+     *   "deletedonlive" => "Deleted"
+     *
+     * Example (with optional title attribute):
+     *   "deletedonlive" => ['text' => "Deleted", 'title' => 'This page has been deleted']
+     */
+    public function getStatusFlags(bool $cached = true): array
```

Comment:
> This seems ... kind of out of scope?

The method is on `SiteTree` right now. It gets called from `CMSMain`, so it needs to be available on any `DataObject` with the `Hierarchy` extension.

I thought of moving it to `DataObject` but decided to put it here instead because it could be useful for models that don't live in the database as well. If I put it on `DataObject` it'd likely end up being moved here anyway when we implement https://github.com/silverstripe/silverstripe-framework/issues/11385 ("Move logic out of DataObject that isn't needed for ORM interactions.")

> I don't like the idea of adding generic functionality that we might use in the future, but we're not actually planning on implementing straight away

It's not being added, just moved. It's already on `SiteTree`. This just makes it available on other models as well.
It _does_ get used right away for other versioned data objects. We're not likely to ever use it in core on arbitrary `ModelData` classes but this allows projects to do so.

> Also the method name for this is wrong for a generic method as 'status flags' is clearly something for versioning

This isn't just for versioning, e.g. fluent uses it: https://github.com/tractorcow-farm/silverstripe-fluent/blob/208b365b66005ce263dbc7ae402fe33885b26d3d/src/Extension/FluentFilteredExtension.php#L61

The method name is also not changing because this is just _moving the method_ not adding a new one. Renaming the methods from `SiteTree` is not the intention here, we're just moving them to somewhere `CMSMain` can rely on them existing.

Reply:
All of that said:
It either has to be here or on `DataObject`.

I can move it to `DataObject` if you prefer just to get us moving along, but since it's not got anything to do with the ORM it doesn't really belong there.

Reply:
> I don't like the idea of adding generic functionality that we might use in the future, but we're not actually planning on implementing straight away

If it makes you happier about this, I'd be okay with widening the scope of the change in https://github.com/silverstripe/silverstripe-versioned/pull/479 to move the call to `statusFlags()` into framework, so all gridfields use it right away.

Reply:
> Does the fluent example currently only apply the status flags on SiteTree and not other DataObjects

Without this PR, yes, because this method only exists on `SiteTree` right now. This PR will automatically make it apply to all `DataObject` models which have the fluent extension applied, since it's the same extension hook - but the flags won't be displayed outside of a `CMSMain` unless the model is also versioned.

> If you did all of the above, would it result in some sort of UX change for regular DataObjects in gridfields? Or is it just adding an essentially unused cache?

I'm not sure what "all of the above" is...  so to cover my bases:
- if this PR is merged, any module or project can apply any status flag to any `ModelData` class - but they'll only be displayed in a `CMSMain` or if the model is versioned.
- if I widen the change in https://github.com/silverstripe/silverstripe-versioned/pull/479 all models will have status flags displayed in gridfields. Right now only versioned models will have their flags displayed in gridfields because it's a versioned extension that adds that display.

Reply:
OK sounds good, so if we widen it would it "just work" from a UX point of view? i.e. would the status flag icons now show on arbitary versioned DataObjects in a GridField with little need for us to modify things?

Reply:
To check that I'll have to implement the change, and at that point we may as well just include that in the PRs. Do you want me to do that?

Reply:
Yeah go for it